MBS-10416: Remove invalid characters from edited annotations

yvanzo opened the pull request:
<!--
    Hello! Thanks for submitting a pull request to MusicBrainz Server.
    We appreciate your time and interest in helping our project!

    Please use this template to help us review your change.

    Depending on your change, some sections may be unneeded, just remove these.
    For example, small pull requests usually don’t need the section “Action”.

    Remember that the more helpful info your pull request includes,
    the easier it is for us to understand and review your changes.

    Ensure that you’ve read through and followed the Contributing Guidelines, at
    https://github.com/metabrainz/musicbrainz-server/blob/master/CONTRIBUTING.md
-->

# Problem
<!--
    Anything that helps us understand why you are making this change goes here.
    What problem are you trying to fix? What does this change address?
-->

MBS-10416: Annotation with some control characters cannot be indexed for search


# Solution
<!--
    Talk about technical details, considerations, or other interesting points.
-->

Strip such characters (and other invalid characters) when entered in annotation.


# Checklist for author
<!--
    The tasks you have to do to get your change ready for review. Use this if
    you draft a pull request. Mark done tasks with an [x] as you progress. See
    https://github.blog/2019-02-14-introducing-draft-pull-requests/
-->

* [x] Add test


# Action
<!--
    Other than merging your change, do you want / need us to do anything else
    with your change? This could include reviewing a specific part of your PR.
-->

1. Clean the 216 existing annotations having such invalid characters.

reosarevok commented on lib/MusicBrainz/Server/Data/Utils.pm:
```diff
@@ -364,6 +364,8 @@ sub trim_comment {
     my $t = shift;
 
+    $t = NFC($t);
+    $t = remove_invalid_characters($t);
```
The other place we use this has a comment:
```
    # Before removing invalid characters, convert space control characters
    # into U+0020 (or else they'll be removed).

```
Is this a worry here?

yvanzo replied:
That comment is not accurate enough, it does not give any example of such character. For example, TAB character is not removed, so I’m not sure which “space control characters” would actually benefit from that.

The other place collapses whitespaces, which we cannot do here because of possible indentation.

It seems neither needed nor appropriate to replace unwanted control characters such as bell with whitespace.

mwiencek replied:
It's talking about any character removed by `remove_invalid_characters` -- there's an explicit list in there (which includes TAB, LF, CR).

Were you perhaps testing this in the release editor? Because that goes through `WS::js::Edit`, not `Form::Annotation`. If you test on any `/edit_annotation` page, this patch will indeed cause it to remove TAB, LF, CR, etc.

(This also suggests we'll have to call `trim_text` where annotations are processed in `WS::js::Edit`.)

yvanzo replied:
> there's an explicit list in there (which includes TAB, LF, CR).

This was not clear because [`remove_invalid_characters`](https://github.com/metabrainz/musicbrainz-server/blob/v-2021-03-01/lib/MusicBrainz/Server/Data/Utils.pm#L398-L411) currently starts by explicitly NOT removing TAB, LF, CR with `s/[^\x09\x0A\x0D…]//gor` but it ends with removing them with `s/[…\p{Cc}…]//g`.

This function will be split using commit https://github.com/metabrainz/musicbrainz-server/pull/1944/commits/2bc4060b8f1abd524b95f2d035546273ee0ff119.

> we'll have to call `trim_text` where annotations are processed in `WS::js::Edit`.

Fixed in commit https://github.com/metabrainz/musicbrainz-server/pull/1944/commits/dbea2f2446aad6aebf8e388d6b4025fe3c268247.

It’s a bit an issue that data processors in `WS::js::Edit` are not systematically synced with their `Form` counterparts. What can we do about that, commenting code with reminders?

mwiencek replied:
Comments are probably the best we can do for now until more significant improvements can be made. If I could go back, I'd probably make /ws/js/edit submit everything through FormHandler and just serialize the field errors it returns.

yvanzo replied:
Reported as ticket MBS-11428 and added comments through commit https://github.com/metabrainz/musicbrainz-server/pull/1944/commits/aee75730894e3c0bc096b5f0603b51a881afca31.

reosarevok commented on lib/MusicBrainz/Server/Data/Utils.pm:
```diff
@@ -360,6 +361,16 @@ sub trim_comment {
+    my $t = shift;
+
+    # Not trimming starting spaces to avoid breaking list formatting,
+    # consider trimming again once this uses Markdown
```
"once this uses Markdown" is less clear in here than in Form::Annotation. Should this be called `trim_annotation` and say "once annotations use Markdown"?

yvanzo replied:
Annotation has a specific syntax for now but there is nothing specific to annotation’s syntax in this function.
It could (or should?) also be used for other text fields such as editor’s bio, edit notes.
Eventually, all of these fields will move to Markdown syntax.

mwiencek replied:
The only issue I have is that it's not distinguished from `trim` (which also operates on text). So I wouldn't know what the difference between `trim` and `trim_text` are from the names alone, I'd have to check usages and git history. What about something like `trim_wikitext`?

yvanzo replied:
Renamed as `trim_multiline_text` in commit https://github.com/metabrainz/musicbrainz-server/pull/1944/commits/2b4597fa01406cc296545c4bbbcd0fd6a1c8b647 since it can be used for editor’s biography, edit note, event setlist, description… which are not necessarily wikitext.

reosarevok replied:
Then I'd suggest "consider trimming again once text fields currently using wikitext start using Markdown" or whatnot maybe?

reosarevok replied:
Actually I see you changed this further down the line so it's probably fine :)

yvanzo replied:
Yes, even in Markdown we may want to preserve leading spaces.

yvanzo commented:
Refactored more code, fixed bugs with TAB, CR, LF and release editor, added/updated tests accordingly.
Also made removing spaces from the end of *every* line through commit https://github.com/metabrainz/musicbrainz-server/pull/1944/commits/9dca2172d2bf5e1a524408a568bc827815a0de33. (That can be added to MBS-10416’s description.)

mwiencek commented:
> “Other, surrogate” characters were/are actually not removed.

Oops, that "Other, surrogate" comment was misleading 'cause they weren't removed below it. They are actually removed by the "trim XML-invalid characters" regular expression (which is good, because they aren't valid UTF-8).

mwiencek reviewed (approved):
LGTM but I'd clarify the commit message about surrogate chars.

mwiencek commented on lib/MusicBrainz/Server/Data/Utils.pm:
```diff
@@ -400,14 +415,19 @@ sub remove_invalid_characters {
     # - Other, private use
     # - Noncharacters
-    =~ s/[\x{200B}\x{00AD}\x{FEFF}\p{Cc}\p{Co}${noncharacter_pattern}]//gr
+    =~ s/[\x{FEFF}\p{Co}${noncharacter_pattern}]//gr
```
I was going to suggest removing `\p{Co}` here to fix MBS-10976, but I submitted that separately instead (https://github.com/metabrainz/musicbrainz-server/pull/1964) so that I could additionally fix the ISE and add tests.

yvanzo commented:
Thanks, I clarified commit message and commented code about surrogate chararacters in commit https://github.com/metabrainz/musicbrainz-server/pull/1944/commits/374ae51801b7dbebb34af05d1ffcd49876afb007.

yvanzo commented:
@brainzbot retest this please

yvanzo commented:
@brainzbot retest this please

yvanzo commented:
Rebased on `master` and resolved conflicts with merged #1964.

reosarevok commented on t/lib/t/MusicBrainz/Server/Controller/Work/AddAnnotation.pm:
```diff
@@ -33,7 +33,7 @@ is_deeply(
             name => 'Dancing Queen'
         },
-        text      => 'Test annotation 6. This is my annotation',
+		    text => "    * Test annotation for a work\r\n    * This annotation has two bullets",
```
Unintended extra spacing?

yvanzo replied:
Thanks, these were unintended tab chars. Removed from commit https://github.com/metabrainz/musicbrainz-server/pull/1944/commits/5b36a92215f127602051b90b500827b6c269b9b7.

reosarevok commented on lib/MusicBrainz/Server/Data/Utils.pm:
```diff
@@ -397,17 +398,23 @@ my $noncharacter_pattern = '\x{FDD0}-\x{FDEF}\x{FFFE}\x{FFFF}';
+
+sub remove_lineformatting_characters {
+    shift
+    # trim other undesirable characters:
```
I'd suggest making this comment more clear - I expect this is not "other undesirable characters" but "line formatting characters"? :)

yvanzo replied:
Specified “lasting line-formatting characters” in commit https://github.com/metabrainz/musicbrainz-server/pull/1944/commits/62348af4bfaa6d665ef5a76e7d945c53199ea37f because some have been removed already.
